src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsRegistryServiceImpl.java

```diff
@@ -1,34 +1,63 @@
 package it.gov.pagopa.pu.classification.service.assessments;
 
-import it.gov.pagopa.pu.classification.dto.generated.CreateAssessmentsRegistryRequest;
+import it.gov.pagopa.pu.classification.connector.debtposition.DebtPositionTypeOrgService;
+import it.gov.pagopa.pu.classification.dto.generated.CreateAssessmentsRegistryByDebtPositionDTOAndIudRequest;
 import it.gov.pagopa.pu.classification.repository.AssessmentsRegistryRepository;
+import it.gov.pagopa.pu.classification.service.BalanceUnmashallerService;
 import it.gov.pagopa.pu.classification.util.SecurityUtils;
 import it.gov.pagopa.pu.classification.util.Utilities;
+import it.gov.pagopa.pu.debtposition.dto.generated.DebtPositionTypeOrg;
+import it.veneto.regione.schemas._2012.pagamenti.ente.CtBilancio;
+import it.veneto.regione.schemas._2012.pagamenti.ente.CtCapitolo;
+import jakarta.transaction.Transactional;
+import java.util.List;
 import org.springframework.stereotype.Service;
 
 @Service
 public class AssessmentsRegistryServiceImpl implements AssessmentsRegistryService{
 
   private final AssessmentsRegistryRepository assessmentsRegistryRepository;
+  private final BalanceUnmashallerService balanceUnmashallerService;
+  private final DebtPositionTypeOrgService debtPositionTypeOrgService;
 
   public AssessmentsRegistryServiceImpl(
-    AssessmentsRegistryRepository assessmentsRegistryRepository) {
+    AssessmentsRegistryRepository assessmentsRegistryRepository,
+    BalanceUnmashallerService balanceUnmashallerService,
+    DebtPositionTypeOrgService debtPositionTypeOrgService) {
     this.assessmentsRegistryRepository = assessmentsRegistryRepository;
+    this.balanceUnmashallerService = balanceUnmashallerService;
+    this.debtPositionTypeOrgService = debtPositionTypeOrgService;
   }
 
+  @Transactional
   @Override
-  public Long createAssessmentsRegistry(CreateAssessmentsRegistryRequest request) {
-    return assessmentsRegistryRepository.insertIfNotExists(
-      request.getOrganizationId(),
-      request.getDebtPositionTypeOrgCode(),
-      request.getSectionCode(),
-      request.getSectionDescription(),
-      request.getOfficeCode(),
-      request.getOfficeDescription(),
-      request.getAssessmentCode(),
-      request.getAssessmentDescription(),
-      request.getOperatingYear(),
-      SecurityUtils.getCurrentUserExternalId(),
-      Utilities.getTraceId());
+  public void createAssessmentsRegistryByDebtPositionDTOAndIud(
+    CreateAssessmentsRegistryByDebtPositionDTOAndIudRequest request, String accessToken) {
+    request.getDebtPositionDTO().getPaymentOptions().stream()
+      .flatMap(paymentOptionDTO -> paymentOptionDTO.getInstallments().stream())
+      .filter(installmentDTO -> request.getIudList().contains(installmentDTO.getIud()))
+      .forEach(i -> {
+        DebtPositionTypeOrg debtPositionTypeOrg = debtPositionTypeOrgService
+          .getDebtPositionTypeOrgByInstallmentId(i.getInstallmentId(), accessToken);
```

Comment:
let's search by id using request.getDebtPositionDTO().getDebtPositionTypeOrgId() instead:
this will be the same for the entire request, so you can fetch it outside this loop